File: express/src/routes/Marketplace/order_items.ts
import { Router } from 'express';
import { prisma } from '../../lib/prisma.js';

const router = Router();

function validate(data: any): void {
  if (!((data.quantity == null || data.quantity > 0))) throw new Error(`Order item quantity must be greater than zero`);
  if (!((data.priceAtPurchase == null || Number(data.priceAtPurchase) >= 0))) throw new Error(`Price at purchase must not be negative`);
}

router.get('/', async (_req, res) => {
  const items = await prisma.orderItem.findMany();
  res.json(items);
});

router.post('/', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.quantity !== undefined) data.quantity = body.quantity;
    if (body.priceAtPurchase !== undefined) data.priceAtPurchase = body.priceAtPurchase;
    if (body.foil !== undefined) data.foil = body.foil;
    if (body.orderId !== undefined) data.orderId = body.orderId;
    if (body.productId !== undefined) data.productId = body.productId;
  try {
  validate(data);
    const entity = await prisma.orderItem.create({ data });
    res.status(201).json(entity);
  } catch (err: any) {
    res.status(400).json({ error: err?.message ?? 'Validation error' });
  }
});

router.get('/:id', async (req, res) => {
  const entity = await prisma.orderItem.findUnique({ where: { id: Number(req.params.id) } });
  if (!entity) return res.status(404).json({ error: 'Not found' });
  res.json(entity);
});

router.put('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.quantity !== undefined) data.quantity = body.quantity;
    if (body.priceAtPurchase !== undefined) data.priceAtPurchase = body.priceAtPurchase;
    if (body.foil !== undefined) data.foil = body.foil;
    if (body.orderId !== undefined) data.orderId = body.orderId;
    if (body.productId !== undefined) data.productId = body.productId;
  try {
  validate(data);
    const entity = await prisma.orderItem.update({ where: { id: Number(req.params.id) }, data });
    res.json(entity);
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : 400;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.quantity !== undefined) data.quantity = body.quantity;
    if (body.priceAtPurchase !== undefined) data.priceAtPurchase = body.priceAtPurchase;
    if (body.foil !== undefined) data.foil = body.foil;
    if (body.orderId !== undefined) data.orderId = body.orderId;
    if (body.productId !== undefined) data.productId = body.productId;
  try {
  validate(data);
    const entity = await prisma.orderItem.update({ where: { id: Number(req.params.id) }, data });
    res.json(entity);
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : 400;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.delete('/:id', async (req, res) => {
  try {
    await prisma.orderItem.delete({ where: { id: Number(req.params.id) } });
    res.status(204).send();
  } catch {
    res.status(404).json({ error: 'Not found' });
  }
});

export default router;

File: gin/internal/handler/marketplace/order_items.go
package handler_marketplace

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	model "cards_project/internal/model/marketplace"
	"cards_project/internal/handler"
)

type OrderItemHandler struct { db *gorm.DB }

func NewOrderItemHandler(db *gorm.DB) *OrderItemHandler {
	return &OrderItemHandler{db: db}
}

func (h *OrderItemHandler) RegisterRoutes(r gin.IRouter) {
	g := r.Group("/api/order_items")
	g.GET("", h.List)
	g.POST("", h.Create)
	g.GET("/:id", h.Get)
	g.PUT("/:id", h.Update)
	g.PATCH("/:id", h.Patch)
	g.DELETE("/:id", h.Delete)
	g.GET("/:id/api/order-items/{id}/total", h.LineTotal)
}

func (h *OrderItemHandler) List(c *gin.Context) {
	skip, limit := handler.Paginate(c)
	var rows []model.OrderItem
	if err := h.db.Offset(skip).Limit(limit).Find(&rows).Error; err != nil {
		handler.DbError(c, err); return
	}
	out := make([]model.OrderItemResponse, len(rows))
	for i := range rows { out[i] = rows[i].ToResponse() }
	c.JSON(http.StatusOK, out)
}

func (h *OrderItemHandler) Create(c *gin.Context) {
	var req model.OrderItemCreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	if msgs := validateOrderItem(&req); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	row := model.OrderItem{}
	row.Quantity = req.Quantity
	row.PriceAtPurchase = req.PriceAtPurchase
	row.Foil = req.Foil
	row.OrderID = req.OrderID
	row.ProductID = req.ProductID
	if err := h.db.Create(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusCreated, row.ToResponse())
}

func (h *OrderItemHandler) Get(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.OrderItem
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "OrderItem"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *OrderItemHandler) Update(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.OrderItem
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "OrderItem"); return }
		handler.DbError(c, err); return
	}
	var req model.OrderItemUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	row.ApplyUpdate(req)
	createReq := toCreateRequestOrderItem(&row)
	if msgs := validateOrderItem(&createReq); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *OrderItemHandler) Patch(c *gin.Context) { h.Update(c) }

func (h *OrderItemHandler) Delete(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	if err := h.db.Delete(&model.OrderItem{}, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "OrderItem"); return }
		handler.DbError(c, err); return
	}
	c.Status(http.StatusNoContent)
}

func (h *OrderItemHandler) LineTotal(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.OrderItem
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "OrderItem"); return }
		handler.DbError(c, err); return
	}
	result, err := row.LineTotal()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.JSON(http.StatusOK, gin.H{"result": result})
}

func validateOrderItem(req *model.OrderItemCreateRequest) []string {
	var errs []string
	if !(req.Quantity > 0) {
		errs = append(errs, "Order item quantity must be greater than zero")
	}
	if !(float64(req.PriceAtPurchase) >= 0) {
		errs = append(errs, "Price at purchase must not be negative")
	}
	return errs
}

func toCreateRequestOrderItem(m *model.OrderItem) model.OrderItemCreateRequest {
	return model.OrderItemCreateRequest{
		Quantity: m.Quantity,
		PriceAtPurchase: m.PriceAtPurchase,
		Foil: m.Foil,
		OrderID: m.OrderID,
		ProductID: m.ProductID,
	}
}
